numpy/core/function_base.py

```diff
@@ -206,3 +206,105 @@ def logspace(start, stop, num=50, endpoint=True, base=10.0, dtype=None):
+        Default is True.
+    dtype : dtype
+        The type of the output array.  If `dtype` is not given, infer the data
+        type from the other input arguments.
+
+    Returns
+    -------
+    samples : ndarray
+        `num` samples, equally spaced on a log scale.
+
+    See Also
+    --------
+    logspace : Similar to geomspace, but with endpoints specified using log
+               and base.
+    linspace : Similar to geomspace, but with arithmetic instead of geometric
+               progression.
+    arange : Similar to linspace, with the step size specified instead of the
+             number of samples.
+
+    Notes
+    -----
+    If the inputs or dtype are complex, the output will follow a logarithmic
+    spiral in the complex plane.  (There are an infinite number of spirals
+    passing through two points; the output will follow the shortest such path.)
+
+    Examples
+    --------
+    >>> np.geomspace(100, 1000, num=4)
+    array([  100.        ,   215.443469  ,   464.15888336,  1000.        ])
+    >>> np.geomspace(100, 1000, num=3, endpoint=False)
+    array([ 100.        ,  215.443469  ,  464.15888336])
+    >>> np.geomspace(100, 1000, num=4, endpoint=False)
+    array([ 100.        ,  177.827941  ,  316.22776602,  562.34132519])
+    >>> np.geomspace(1, 256, num=9)
+    array([   1.,    2.,    4.,    8.,   16.,   32.,   64.,  128.,  256.])
+
+    Graphical illustration:
+
+    >>> import matplotlib.pyplot as plt
+    >>> N = 10
+    >>> x1 = np.geomspace(1, 10, N, endpoint=True)
+    >>> x2 = np.geomspace(1, 10, N, endpoint=False)
+    >>> y = np.zeros(N)
+    >>> plt.plot(x1, y, 'o')
+    [<matplotlib.lines.Line2D object at 0x...>]
+    >>> plt.plot(x2, y + 0.5, 'o')
+    [<matplotlib.lines.Line2D object at 0x...>]
+    >>> plt.ylim([-0.5, 1])
+    (-0.5, 1)
+    >>> plt.show()
+
+    """
+    if start == 0 or stop == 0:
+        raise ValueError('Geometric sequence cannot include zero')
+
+    dt = result_type(start, stop, float(num))
+    if dtype is None:
+        dtype = dt
+    else:
+        # complex to dtype('complex128'), for instance
+        dtype = _nx.dtype(dtype)
+
+    # Avoid negligible real or imaginary parts in output by rotating to
+    # positive real, calculating, then undoing rotation
+    out_sign = 1
+    if _nx.real(start) == _nx.real(stop) == 0:
+        start, stop = start.imag, stop.imag
+        out_sign = 1j * out_sign
+    if _nx.sign(start) == _nx.sign(stop) == -1:
+        start, stop = -start, -stop
+        out_sign = -out_sign
+
+    # Promote both arguments to the same dtype in case, for instance, one is
+    # complex and another is negative and log would produce NaN otherwise
+    log_start = _nx.log10(dtype.type(start))
+    log_stop = _nx.log10(dtype.type(stop))
+    result = out_sign * logspace(log_start, log_stop, num=num,
+                                 endpoint=endpoint, base=10.0, dtype=dtype)
+
+    return result.astype(dtype)
```

Comment:
is this needed?  could probably just return on the above line, since dtype will be handled by logspace and out_sign?